### ISSUE TITLE
ECU-B time out solution

### ISSUE THREAD

**Issue by Xqlus1ve91** [ksheumaker/homeassistant-apsystems_ecur#179]
Currently running the following:

V1.2.29
ECU-B, firmware 2163xxxxxx

i noticed now for the second time that the data in HA stops and keeps running at the last know value, i then need to delete the integration and wait for an hour or-so for me to get it back to work.

what is the best solution to prevent this from happening? Do i need to restart my ecu every day or 12 hours?

**Comment by HAEdwin** [comment]
I often read that the integration is uninstalled and reinstalled, that makes absolutely no sense. Software and settings will remain exactly the same.

Because you write that you sometimes have to wait an hour or-so before new data appears, you should also check whether the Zigbee connection between the ECU and Inverter(s) is still established. You can check this by looking at the return of energy. 
If you have a smart electricity meter you should still see an energy return, then the Zigbee signal between the ECU and the inverter has been lost. It is then sufficient to move the ECU or reorient the antennas, you then have to wait a maximum of 5 minutes for the next signal to arrive. The Zigbee signal can be (temporarily) disrupted by other sources or weather conditions. 

In summary, you can pay attention to the following if you do not receive any new data:
- Zigbee signal is too weak
- ECU is stuck (see log) restart the ECU (ECU-C old model or ECU-B)
- The ECU query device switch is still in the off position (will be reset if you reinstall but you can also turn it back on)
- Inverters are offline

**Comment by Xqlus1ve91** [comment]
The data from the ecu towards the cloud works fine and is detailed per 5 minutes. My connection between ecu and repeater i should say is fine and strong enough for good reception. So when i have these issues in HA and i look in the EMA app everything is ok and reporting. Inverters are not offline 

An Query switch, where to look for and how to implement?is there a instruction for that? In the mean time i will check issue #163.

**Comment by HAEdwin** [comment]
How often do you query the ECU? Data on EMA might have been made complete in de evenings, check at the moment the problems occur. The switch.ecu_query_device is an entity you can add to the dashboard, check it's status. The ECU-B should be working just as stable as the ECU-R.

**Comment by HAEdwin** [comment]
Did you check the home-assistant.log?

**Comment by Xqlus1ve91** [comment]
Yes i have, i have seen that the query device has been shut off and that that causes the data to stop updating. I have put the switch on my dashboard in case it happens again, hopefully switching it back on get my readings back alive.

Second thing i noticed is that the 'ECU using cache data' enitity keeps turning on and off,  is normal behavior?

![45678](https://user-images.githubusercontent.com/121225361/233697382-ad5cbaeb-7a22-45b8-a4a1-c82b85b5fd6c.PNG)

**Comment by HAEdwin** [comment]
The frequency in which the cache is being used is way too often. Check the quality of WiFi connection and whether there is any interference. If the integration fails to do the query the set number of times in a row, it is assumed that the ECU needs a reset. The switch ensures that you can be alerted to, for example, your telephone via an automation. But again the cache is being used too often in your case. I've had it twice during the last 24 hours (ECU-R) and that's normal behaviour. It also could be a firmware issue.

**Comment by Xqlus1ve91** [comment]
Well, we are a couple of days further, my caching switching has minimised to only once a day, i haven’t changed anything so far. But this was only happening on days where i got maximum energy production so i will keep a close eye to another day like those.

**Comment by HAEdwin** [comment]
There also seems to be an issue with HA installations on Intel NUC's. I'm using an ODROID N2+ and works just as stable as a RPi.

**Comment by Xqlus1ve91** [comment]
I am not using a Intel NUC, I am using a Intel CPU on a Synology NAS in VM. The issue still persists, i am currently trying to configure an automation that will switch my ECU on and off if the 'using cache data' has turned ON for 3 or 5 times. Still struggling how to create a proper counter for it that will activate the automation an afterwards reset the counter.

**Comment by HAEdwin** [comment]
Yes, but I can also imagine that communication was lost somehow, might be signal interference from a device nearby if it happens at 14.00 every day.

**Comment by HAEdwin** [comment]
@P-Verbrugge @Xqlus1ve91 It is acceptable if the ECU does not respond to the query at times, but the next query usually succeeds. If all else fails, an (automated) power cycle from the ECU always works. I'll close this issue for now, If there is reason to do so, feel free to reopen the issue or create a new issue.
